<a href="https://colab.research.google.com/github/agcbls49/THESIS/blob/main/THESIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INTRODUCTION**
This colab file contains the data preprocessing explanations, model training, and model comparisons of YOLOv5n, YOLOv8n, YOLOv11n and YOLOv26n to be utilized in the thesis entitled "Vocasee: Voice-Guided Indoor Object Finder Android Application for Visually Impaired Users".

<br>

The dataset image annotations and this code was made by Amazing Grace O. Cabiles.


---



## **DATA PREPROCESSING**
This step was done on a local machine but the code for all data preprocessing related tasks are uploaded in Github. You can visit this Github [repository](https://github.com/agcbls49/THESIS) and look through the code and the README file that contains the short explanations of its contents.


---



## **SETUP**

### Mounting the Dataset from Google Drive to be used

This step mounts the Google Drive where the dataset is stored and copies the dataset.zip file to Google Colab's storage allowing the dataset to be used for model training.

In [4]:
from google.colab import drive
drive.mount('/content/gdrive')

# Copy dataset.zip from Drive to Colab's local storage
!cp "/content/gdrive/MyDrive/2_THESIS_DATASET/dataset.zip" /content/

### Unzipping the dataset contents
This step simply unzips the content of dataset.zip in Google Colab's storage.

In [5]:
# Unzip the dataset folder contents
!unzip -q /content/dataset.zip -d /content/dataset

### Installing Ultralytics for YOLO
This step installs the ultralytics package which is needed for training YOLO models.

In [6]:
# Install ultralytics  for YOLO
!pip install ultralytics -q

### Creating a YAML file
This step creates a YAML file which is needed for training YOLO models. It contains the path of the unzipped dataset, where the train, validation and test images are located, and the names of the classes along with their class numbers.

In [15]:
import yaml
import os

data = {
    'path': '/content/dataset/dataset',
    'train': 'images/train',
    'val': 'images/val',
    'test': 'images/test',
    'nc': 20,
    'names': [
        'backpack',
        'book',
        'cup',
        'glasses',
        'headphone',
        'key',
        'phone',
        'remote',
        'wallet',
        'water bottle',
        'body spray',
        'card',
        'charger',
        'comb',
        'flashlight',
        'glasses case',
        'medicine',
        'nail clipper',
        'shoe',
        'watch'
    ]
}

yaml_path = '/content/data.yaml'

# write file
with open(yaml_path, 'w') as f:
    yaml.dump(data, f)

print("written")

# verify file exists
print("exists:", os.path.exists(yaml_path))

# verify size
print("size:", os.path.getsize(yaml_path))

# print contents
with open(yaml_path, 'r') as f:
    content = f.read()

print("CONTENT:")
print(content)

written
exists: True
size: 298
CONTENT:
names:
- backpack
- book
- cup
- glasses
- headphone
- key
- phone
- remote
- wallet
- water bottle
- body spray
- card
- charger
- comb
- flashlight
- glasses case
- medicine
- nail clipper
- shoe
- watch
nc: 20
path: /content/dataset/dataset
test: images/test
train: images/train
val: images/val



# **SPECIFICATIONS**

The dataset is split into **80% train, 10% valid and 10% test** and every model is trained using the following specifications:
1. 50 Epochs
2. All images resized to 640x640
3. 16 Batches

# **DELETE CACHE**
The code deletes the cached yaml file ensuring that the code runs the proper yaml file.

In [17]:
!find /content/dataset/dataset -name "*.cache" -delete

## **YOLO v5 NANO MODEL**

In [18]:
!yolo detect train data=/content/data.yaml model=yolov5nu.pt epochs=50 imgsz=640 batch=16 name=yolov5n_run

Ultralytics 8.4.50 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov5nu.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov5n_run-5, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, p

### **YOLO v5 NANO RESULTS**
Precision = 0.98137 <br>
Recall = 0.92967 <br>
mAP50 = 0.97396 <br>
mAP50-95 = 0.85258 <br>

---
### **INTERPRETATION**

When the model draws a box, 98.1% of the boxes drawn are correctly matched from the annotated boxes or the ground truth. Out of all the annotated ground truth, 92.9% of them were detected by the model. When the model draws a box and it has at least a 50% overlap with the ground truth box, the average precision across all classes is 97.4%. Lastly, when the model draws a box and the overlap requirement goes from 50% to 95% in 5% steps, the average precision across all classes averaged across all those steps is 85.3%.

---
### **OVERALL**
The model performs very well and it rarely draws wrong boxes (only 1.9% false alarms). It finds most of the ground truth boxes (misses only 7.1%).

# **YOLO v8 NANO MODEL**

In [ ]:
!yolo detect train data=/content/data.yaml model=yolov8n.pt epochs=50 imgsz=640 batch=16 name=yolov8n_run

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics 8.4.46 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, i

### **YOLO v8 NANO RESULTS**
Precision = 0.9855 <br>
Recall = 0.92851 <br>
mAP50 = 0.97397 <br>
mAP50-95 = 0.85881 <br>

---
### **INTERPRETATION**
When the model draws a box, 98.6% of the boxes drawn are correctly matched from the annotated boxes or the ground truth. Out of all the annotated ground truth, 92.9% of them were detected by the model. When the model draws a box and it has at least a 50% overlap with the ground truth box, the average precision across all classes is 97.4%. Lastly, when the model draws a box and the overlap requirement goes from 50% to 95% in 5% steps, the average precision across all classes averaged across all those steps is 86%.

---
### **OVERALL**
The model performs very well and it rarely draws wrong boxes (only 1.4% false alarms). It finds most of the ground truth boxes (misses only 7.1%).
<br><br>
**When compared to YOLOv5n, it seems to perform the same but YOLOv8n has less false alarms.**

# **YOLO v11 NANO MODEL**

In [ ]:
!yolo detect train data=/content/data.yaml model=yolo11n.pt epochs=50 imgsz=640 batch=16 name=yolo11n_run

Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolo11n_run, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, pers

### **YOLO v11 NANO RESULTS**
Precision = 0.97147 <br>
Recall = 0.93703 <br>
mAP50 = 0.97707 <br>
mAP50-95 = 0.8656 <br>

---
### **INTERPRETATION**
When the model draws a box, 97.1% of the boxes drawn are correctly matched from the annotated boxes or the ground truth. Out of all the annotated ground truth, 93.7% of them were detected by the model. When the model draws a box and it has at least a 50% overlap with the ground truth box, the average precision across all classes is 97.7%. Lastly, when the model draws a box and the overlap requirement goes from 50% to 95% in 5% steps, the average precision across all classes averaged across all those steps is 86.6%.

---
### **OVERALL**
The model performs very well and it rarely draws wrong boxes (only 2.9% false alarms). It finds most of the ground truth boxes (misses only 6.3%). <br><br> **When compared to YOLOv5n and YOLOv8n, the performance is slightly lower, although the difference is minimal.**

# **YOLO v26 NANO MODEL**

In [ ]:
!yolo detect train data=/content/data.yaml model=yolo26n.pt epochs=50 imgsz=640 batch=16 name=yolo26n_run

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics 8.4.46 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, i

### **YOLO v26 NANO RESULTS**
Precision = 0.92703 <br>
Recall = 0.94372 <br>
mAP50 = 0.96516 <br>
mAP50-95 = 0.84323 <br>

---
### **INTERPRETATION**
When the model draws a box, 92.7% of the boxes drawn are correctly matched from the annotated boxes or the ground truth. Out of all the annotated ground truth, 94.4% of them were detected by the model. When the model draws a box and it has at least a 50% overlap with the ground truth box, the average precision across all classes is 96.5%. Lastly, when the model draws a box and the overlap requirement goes from 50% to 95% in 5% steps, the average precision across all classes averaged across all those steps is 84.3%.

---
### **OVERALL**
The model performs very well and it rarely draws wrong boxes (only 7.3% false alarms). It finds most of the ground truth boxes (misses only 5.6%). <br><br> **When compared to the previous YOLO models, the performance is slightly lower overall, although the difference is minimal. The model prioritizes detecting more objects (higher recall) but is less precise, resulting in more false alarms**

# **EXPORTING THE BEST MODEL**

The code below copies the best trained model and saves it on the cloud or Google Drive.

In [ ]:
!cp /content/runs/detect/yolo11n_run/weights/best.pt "/content/gdrive/MyDrive/yolo11n_best.pt"
print("Saved to Drive!")

Saved to Drive!


This code exports the trained model as a Tensor Flow Lite file allowing it to be deploying on mobile devices.

In [ ]:
!yolo export model=/content/runs/detect/yolo11n_run/weights/best.pt format=tflite imgsz=640 int8=False

Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CPU (Intel Xeon CPU @ 2.00GHz)
💡 ProTip: Export to OpenVINO format for best performance on Intel hardware. Learn more at https://docs.ultralytics.com/integrations/openvino/
YOLO11n summary (fused): 101 layers, 2,584,102 parameters, 0 gradients, 6.3 GFLOPs

PyTorch: starting from '/content/runs/detect/yolo11n_run/weights/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 14, 8400) (5.2 MB)
requirements: Ultralytics requirements ['onnx>=1.12.0,<2.0.0', 'onnxslim>=0.1.71', 'onnxruntime'] not found, attempting AutoUpdate...
Using Python 3.12.13 environment at: /usr
Resolved 12 packages in 312ms
Prepared 4 packages in 1.52s
Installed 4 packages in 287ms
 + colorama==0.4.6
 + onnx==1.21.0
 + onnxruntime==1.26.0
 + onnxslim==0.1.93

requirements: AutoUpdate success ✅ 2.9s
WARNING ⚠️ requirements: Restart runtime or rerun command for updates to take effect


ONNX: starting export with onnx 1.21.0 opset 20...
ONNX: slim

This code downloads the .tflite to the local machine.

In [ ]:
from google.colab import files
import glob

tflite_files = glob.glob('/content/runs/detect/yolo11n_run/weights/**/*.tflite', recursive=True)
if tflite_files:
    files.download(tflite_files[0])
else:
    print("No tflite found — check Cell 8 output for the path")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# **COMPARISON TABLE**


| Model Name | Precision  | Recall     | mAP50      | mAP50-95   | False Alarms | Misses     |
| ---------- | ---------- | ---------- | ---------- | ---------- | ------------ | ---------- |
| YOLOv5n    | 98.1%      | 93%        | 97.4%      | 85.3%      | 1.9%         | 7.1%       |
| YOLOv8n    | 98.6%      | 92.9%      | 97.4%      | 86%        | 1.4%         | 7.1%       |
| YOLOv11n   | 97.1%      | 93.7%      | 97.7%      | 86.6%      | 2.9%         | 6.3%       |
| YOLOv26n   | 92.7%      | 94.4%      | 96.5%      | 84.3%      | 7.3%         | 5.6%       |




---
## **TABLE INTERPRETATION**

All four models demonstrate strong performance, but they differ in how they balance precision, recall, and overall detection quality. **YOLOv8n** achieves the highest precision (98.6%) and the lowest false alarm rate (1.4%), making it the most reliable in avoiding incorrect detections, although it has slightly lower recall (92.9%) and misses the same proportion of objects as YOLOv5n. **YOLOv5n** provides a very balanced performance, with high precision (98.1%), solid recall (93%), and competitive mAP scores, making it a stable and consistent option overall. **YOLOv11n** stands out by achieving the highest mAP50 (97.7%) and mAP50–95 (86.6%), indicating the best overall detection quality and bounding box accuracy; it also improves recall (93.7%) compared to v5n and v8n, resulting in fewer missed objects, though at the cost of slightly higher false alarms (2.9%). In contrast, **YOLOv26n** prioritizes recall (94.4%), achieving the fewest misses (5.6%), but this comes with a significant drop in precision (92.7%) and the highest false alarm rate (7.3%), as well as the lowest mAP scores, indicating weaker overall accuracy.



---

# **CONCLUSION**

Considering all metrics, **YOLOv11n is the best overall model**, as it provides the highest detection accuracy (mAP), a strong balance between precision and recall, and improved object coverage with only a minimal increase in false positives.


